torch/_dynamo/output_graph.py

```diff
@@ -255,26 +255,38 @@ def __init__(
             "co_firstlineno": f_code.co_firstlineno,
         }
 
+        # tracked_fakes says where any tensor that was wrapped to fake came
+        # from.  It is similar to GraphArg, in that all GraphArgs will get
+        # will get added to TrackedFakes, but TrackedFakes also contains
+        # GraphArgs that got pruned, and things like Tensor attributes which
+        # aren't explicit graph inputs.  Used by shape guard
+        self.tracked_fakes: List[TrackedFake] = []
+
+        shape_env = ShapeEnv(
+            # Reference Cycle!
+            # Share a reference to the list of TrackedFake.
+            #
+            # ShapeEnv needs this in order to be able to reproduce the call
+            # to produce_guards at an arbitrary time point. That is because
+            # TrackedFake instances may have its metadata changed throughout
+            # the program execution.
+            tracked_fakes=self.tracked_fakes,
```

Comment:
Calling out reference cycle `OutputGraph` <-> `ShapeEnv`.

Reply:
sorry, how do you get from ShapeEnv to OutputGraph?